ql/src/test/org/apache/hadoop/hive/ql/optimizer/calcite/stats/TestFilterSelectivityEstimator.java

```diff
@@ -508,8 +492,19 @@ public void testComputeRangePredicateSelectivityBetweenWithNULLS() {
   public void testComputeRangePredicateSelectivityNotBetweenWithNULLS() {
     doReturn((double) 20).when(tableMock).getRowCount();
     doReturn(Collections.singletonList(stats)).when(tableMock).getColStat(Collections.singletonList(0));
-    RexNode filter = REX_BUILDER.makeCall(SqlStdOperatorTable.BETWEEN, boolTrue, inputRef0, int1, int3);
+    RexNode filter = makeNotBetween(inputRef0, int1, int3);
     FilterSelectivityEstimator estimator = new FilterSelectivityEstimator(scan, mq);
     Assert.assertEquals(0.55, estimator.estimateSelectivity(filter), DELTA);
   }
+
+  private RexNode makeNotBetween(RexNode inputRef, RexNode left, RexNode right) {
+    RexNode withOr = REX_BUILDER.makeCall(
+        SqlStdOperatorTable.OR,
+        REX_BUILDER.makeCall(SqlStdOperatorTable.LESS_THAN, inputRef, left),
+        REX_BUILDER.makeCall(SqlStdOperatorTable.GREATER_THAN, inputRef, right)
+    );
+    RexSimplify simplify = new RexSimplify(REX_BUILDER, RelOptPredicateList.EMPTY, RexUtil.EXECUTOR);
+
+    return simplify.simplify(withOr);
```

Comment:
how about 
```
return REX_BUILDER.makeCall(SqlStdOperatorTable.NOT, REX_BUILDER.makeBetween(inputRef0, int1, int3));
```

Reply:
What happens if somehow we end up with an expression like
```
NOT(SEARCH($0, Sarg[[<lower>...<upper>]]))
```
in one of the query plans ?

Maybe I'm missing something but in case of `NOT(<expr_a>)` I would expect a recursive call something like
```
case NOT:
  return 1 - estimate(<expr_a>);
```
I found that handling `NOT` was added by https://issues.apache.org/jira/browse/HIVE-8768 but description is missing.

Reply:
Unfortunately, for NOT, we instead call the method `computeNotEqualitySelectivity` and it returns incorrect results, as described above. Maybe we need to check if NOT is implemented correctly, however, I am not very knowledgeable with the theory of selectivity calculations.

I have noticed through tests that NOT BETWEEN never adds a NOT in front of the SEARCH operator, like you have described above. It instead includes multiple ranges in the Sarg. However, we should definitely ensure that we handle NOT(SEARCH(...)) correctly in Hive as it is possible make such a RexNode. I will try to look into this, however, if you have some time, could you also look into this please?

Reply:
I also noticed that both `BETWEEN` and `NOT BETWEEN` are currently represented with type `SqlKind.BETWEEN` and differentiated by a boolean. So `NOT BETWEEN` doesn't go through `case NOT:` in `FilterSelectivityEstimator`, it goes through `case BETWEEN:`.

I think we might need to change how things are handled in `NOT` if we want to handle `NOT(SEARCH(...))`.

Given that we fall through `case NOT` into `case NOT_EQUALS`, I suspect if we only support `NOT` as equivalent to `NOT_EQUALS`, and not for chained operators like `NOT(IN(...))` or `NOT(BETWEEN(...))`.
```
    case NOT:
    case NOT_EQUALS: {
      selectivity = computeNotEqualitySelectivity(call);
      break;
    }
```

Reply:
In Calcite 1.33, `NOT SEARCH...` is simplified to `SEARCH`: https://github.com/apache/calcite/blob/calcite-1.33.0/core/src/main/java/org/apache/calcite/rex/RexSimplify.java#L741

I have made the change that you suggested - using `NOT` and `makeBetween` to create `NOT BETWEEN`, however, I also had to remove a test where right value was lower than left, as it throws an exception.